cpp/roaring64map.hh

```diff
@@ -0,0 +1,796 @@
+     * bitmap,
+     * writing the result in the current bitmap. The provided bitmap is not
+     * modified.
+     */
+    Roaring64Map &operator^=(const Roaring64Map &r) {
+        for (const auto& map_entry : r.roarings) {
+            if (roarings.count(map_entry.first) == 0) {
+                roarings[map_entry.first] = map_entry.second;
+                roarings[map_entry.first].setCopyOnWrite(copyOnWrite);
+            }
+            else
+                roarings[map_entry.first] ^= map_entry.second;
+        }
+        return *this;
+    }
+
+    /**
+     * Exchange the content of this bitmap with another.
+     */
+    void swap(Roaring64Map &r) { roarings.swap(r.roarings); }
+
+    /**
+     * Get the cardinality of the bitmap (number of elements).
+     */
+    uint64_t cardinality() const {
+        return std::accumulate(roarings.cbegin(), roarings.cend(), 0,
+            [](uint64_t previous, const std::pair<uint32_t, Roaring>& map_entry) {
+                return previous + map_entry.second.cardinality();
+            });
+    }
+
+    /**
+    * Returns true if the bitmap is empty (cardinality is zero).
+    */
+    bool isEmpty() const {
+        return std::all_of(roarings.cbegin(), roarings.cend(),
+            [](const std::pair<uint32_t, Roaring>& map_entry) {
+                return map_entry.second.isEmpty();
+            });
+    }
+
+    /**
+    * Returns true if the bitmap is subset of the other.
+    */
+    bool isSubset(const Roaring64Map &r) const {
+        for (const auto& map_entry : roarings) {
+            auto roaring_iter = r.roarings.find(map_entry.first);
+            if (roaring_iter == roarings.cend())
+                return false;
+            else
+                if (!map_entry.second.isSubset(roaring_iter->second))
+                    return false;
+        }
+        return true;
+    }
+
+    /**
+    * Returns true if the bitmap is strict subset of the other.
+    */
+    bool isStrictSubset(const Roaring64Map &r) const { return isSubset(r) && !(*this == r); }
+
+    /**
+     * Convert the bitmap to an array. Write the output to "ans",
+     * caller is responsible to ensure that there is enough memory
+     * allocated
+     * (e.g., ans = new uint32[mybitmap.cardinality()];)
+     */
+    void toUint64Array(uint64_t *ans) const {
+        std::accumulate(roarings.cbegin(), roarings.cend(), ans,
+            [](uint64_t* previous, const std::pair<uint32_t, Roaring>& map_entry) {
+                for (uint32_t low_bits : map_entry.second)
+                    *previous++ = (uint64_t(map_entry.first) << 32) | uint64_t(low_bits);
+                return previous;
+            });
+    }
+
+    /**
+     * Return true if the two bitmaps contain the same elements.
+     */
+    bool operator==(const Roaring64Map &r) const {
+        // we cannot use operator == on the map because either side may contain empty Roaring Bitmaps
+        auto lhs_iter = roarings.cbegin();
```

Comment:
I added a number of comments to improve readability. The maps do not remove empty roaring elements unless shrinkToFit() is called. Thus we need a more complicated equality that skips over any empties found.